### ISSUE TITLE
Absolutely Essential Realtime Masking

### ISSUE THREAD

**Issue by adamearle** [opentoonz/opentoonz#4329]
**Feature Request**
Column masking ability to be turned back on through the code that is available and refactored for speed up. Masking is super important to animation and this feature not being available forces users to work at a much slower pace increasing budgets significantly.

**Solution**
The ability to create a **"mask column"** that is either vector or raster. This should provide real-time feedback for artists when the animator if is not using **"Preview mode"** or **"Sub-Camera Preview mode"** in the viewer or combo viewer.

Matte effects would be welcomed if they were in real-time as well. This is difficult because if blurs and other kinds of FX nodes maybe work together to create a matte. This is why a simple Column Mask that's either raster or vector would be the best solution here.

**Comment by age3rcm** [comment]
![1573409592920](https://user-images.githubusercontent.com/36406840/158353968-fed75e86-cbca-40f9-bb6f-07cac69ab937.png)
PPPPPPLLLLLLLEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEAAAASSSEEEEE

**Comment by F-Burning** [comment]
Very necessary!

**Comment by adamearle** [comment]
@RodneyBaker has pointed me to this report. I still can't help but think that nodes are the wrong direction for masking to happen in real time. 

If the nodes were to do real-time masking, my suggestion would be to make them the same as existing columns except add a checkbox that says mask. This would signify that the current columns would be the mask version of the current column types that already exist.

1. Vector [Masked]
2. Toonz Raster  [Masked]
3. Raster  [Masked]

Implementation of vector masking could be an opportunity to introduce Natron or Blender roto (masking) nodes path system into OpenToonz. At least there would be some code to look at as a launchpad for getting masking started. Something to think about.

Link:
https://github.com/tahoma2d/tahoma2d/pull/361

**Comment by adamearle** [comment]
This could also extend into SVG code https://github.com/opentoonz/opentoonz/issues/4051

**Comment by Bracket-H** [comment]
Column masking? 
Ability to be turned back on again? 

Did Opentoonz have this functionality and it was taken out, or disabled? 

Regardless, what is column masking? Matte? I know what a mask is but what functionality are you wishing for here?

**Comment by gab3d** [comment]
Ultimately the goal should be to have realtime (or at least 'super fast') previewing in OT, so that by default the Preview mode could be ON.

**Comment by F-Burning** [comment]
The problem I see with columns is the space in the Xsheet/timeline, because create columns every time you need a mask maybe could be chaotic. As @gab3d said, the point would be works with nodes in real time instead of previews, but it is such an important change that I think it will never be done, unfortunately.

I have realized that Opentoonz has such a specific way of working, that you can't do anything that goes out of line because it will be problematic.

**Comment by adamearle** [comment]
> 

That's incredibly disheartening. Does this mean we shouldn't try?

As for columns stacking. I am sure users will have no problem with that Adobe After Effects does exactly the same thing, I know that I won't. I would rather have something that I can work within real-time than nothing at all.

I do understand what you are saying. Though in order to attract more development we really need to be getting something down. I agree with you nodes would be great. If that is the fastest and easiest direction to go then I would say,

                                                                                LETS DOOOOO IIIIIT!
                                                                                
If it's not then I honestly don't really care. The idea is masking in real-time. If means the matte in and out nodes are rewritten then kool. I am sure I don't need to tell you OT needs this.

Fixing mask, Fixing keys is going to make life easier. Which means more users. Which means more development. Which means more money. Which means a better tool. I am trusting that the developer will be able to do what they do best and spot the problems. So we can all start working out how to make it happen.

Even if it means passing the hat around getting someone to work on it for 3 to 4 months so it's solid. It has to start somewhere.

**Comment by Bracket-H** [comment]
I still don't understand what everyone wants here. Not trolling.
is this about surprises? like, matte...live matte...is needed because otherwise the user doesn't know what the matte does?

Cause ...you can use something as a matte. Like, if I uploaded a toonz raster level, that has transparency, but then a sphere in the middle, you can see the sphere.
You could expose this level on the X-sheet, and put it on top of a column, and the sphere would show...there.

So, seeing the sphere there, you'd know where the matte is, if you used it as a matte input. It doesn't need a preview.
The preview would only show the effect of the matte, but that would be the same as the matte showing...as a column.
Even a feathered matte would show, since OT would blend the matte column correctly, with the feathered alpha. 
The preview only shows the effect of the matte.

A realtime version would just be that, a preview that is just faster but ...the program still has to compute the effect of the matte/mask (which a preview does) 

A dedicated code region in the program that focuses on that would make it faster but ...yeah I personally don't see the ultimate benefit here.
And it's not me being a contrarian, or anything. I truly just don't know what exactly is being wanted here, and why it would be such a game changer.

I am speaking of ignorance if I offend you, not malice. In other words: I don't get it.
And I use mattes and masks in a lot of things...I use compositors, I write 3D and 2D games. Etc.

But this particular topic ...I don't know what's being wanted. What is a mask column? 
The only thing I can imagine is ...what I just described above...and OT can already do that..so @_@ yeah.

**Comment by shun-iwasawa** [comment]
I'm planning to develop the mask feature in the near future (probably will work on a prototype in V1.7 period).

My blueprint of the feature would be similar to the one in Adobe AfterEffects.
That is to say, the mask shape will be specified by bezier curve (i.e. a vector shape with no line thickness). And it will be more a "property" of the column rather than the "mask column"; a column with special purpose.

Yes, I would like to make the mask to take effect in real-time which will be handy, but I think the major advantage of this feature over the Matte Fx would be that users do not need to save a level file for mask shape since the mask shape (and its animation) will be saved in the scene file.

**Comment by gab3d** [comment]
@shun-iwasawa 
great to know! 
will there be some sort of "feather" and "expansion" features for those masks, as in AE?

**Comment by RodneyBaker** [comment]
Tagging with roadmap as @shun-iwasawa has stated plans to work on related area.
As we get more information on that effort we can assess how closely it addresses the requirements of AERM (TM) as well as that  requested by @konero, @melieconiek and others.

**Comment by adamearle** [comment]
That is an interesting idea as well on the mask @shun-iwasawa. I might be worthwhile  building an option in there to choose:

1. Bezier mask 
2.Choose Colum

Choose Column may come with a warning, "Playback may be slow because of a complex connection made in the schematic view".

You may also want to take a look at this project for some ideas? https://github.com/MaurycyLiebner/enve I spent yesterday afternoon looking at this and pushing it. The only thing that i would say about the "enve" Project is that there can't be new bezier objects from frame to frame. I have outlined that here https://github.com/MaurycyLiebner/enve/issues/302

There are some nice ideas in this project for you to build upon and play with. I hope you like it.

**Comment by konero** [comment]
I only just stumbled on this thread. I don't know how well you guys know Toon Boom Harmony, but in this software they handle this with a node (sort of), it's called the **Animated Matte Generator** and just like Toonz you can filter out specific color indexes you want to attach the animated matte to.

There's a neat video of a user explaining how to use it in detail here: https://youtu.be/g_PyeLb4czg?t=225

**Comment by adamearle** [comment]
@konero yeah I have been a TBH rigger from the way. Making with nodes makes a lot of sense to me. many applications take this approach as well. I think a real-time masking node could be implemented.

**Comment by RodneyBaker** [comment]
This is a note to let everyone know that this general feature of column masking is present in a working release of Tahoma2D.
As some aspects of the feature will need additional work (such as how column masking works with FX) inclusion is not expected for some time in a final release of T2D. 

For reference please see:  https://github.com/tahoma2d/tahoma2d/pull/1142

**Comment by RodneyBaker** [comment]
In addition to vector level masks @manongjohn has added smart raster (toonz raster) and raster levels as column masks.
There is still some work to be done so the Pull Request is WIP for now with target release of T2D v1.5 (Dec 2024 timeframe).

REF:  https://github.com/tahoma2d/tahoma2d/pull/1307

**Comment by konero** [comment]
Hmm I think a node is the preferable way, to answer the worry about real-time playback a good shrink value and caching the nodes before the mask would make it pretty fast. I think it would be easier as there are already FX gadgets that support 'point animation' in the function editor panel, so perhaps we just need a new gadget (that can support double points to calculate the feather, like Natron/Resolve).

I mentioned in another issue, that 'Corridor Gradient Iwa' is pretty close already, has 90% of the functionality. But just needs to support a source input and have the gadget support paths instead of boxes. One way I guess would be to have a extended toolbar popup on the viewer when in preview mode that can be context aware of the node, which would have add point, remove point, make feather point, remove feather point, other bezier functions, etc.) so the user can draw the mask gadget shape.
